lib/jwt/jwt.go

```diff
@@ -797,3 +810,103 @@ func (k *Key) VerifyPluginToken(token string, claims PluginTokenParam) (*Claims,
+	RedirectURI  string `mapstructure:"redirect_uri"`
+	ResponseType string `mapstructure:"response_type"`
+
+	// Optional parameters
+	OptionalParameters map[string]any `mapstructure:"-,remain"`
+}
+
+// Check validates that required parameters are set.
+func (o *OIDCAuthRequestClaims) Check() error {
+	if o.ClientID == "" {
+		return trace.BadParameter("client id missing")
+	}
+	if o.Scope == "" {
+		return trace.BadParameter("scope missing")
+	}
+	if o.RedirectURI == "" {
+		return trace.BadParameter("redirect uri missing")
+	}
+	if o.ResponseType == "" {
+		return trace.BadParameter("response type missing")
+	}
+	return nil
+}
+
+func (o *OIDCAuthRequestClaims) toMap() map[string]any {
+	requestMap := map[string]any{}
+	maps.Copy(requestMap, o.OptionalParameters)
+
+	requestMap["client_id"] = o.ClientID
+	requestMap["scope"] = o.Scope
+	requestMap["redirect_uri"] = o.RedirectURI
+	requestMap["response_type"] = o.ResponseType
+
+	return requestMap
+}
+
+// oidcAuthRequestFromClaims produces an OIDCAuthRequestClaims from a map containing
+// OIDC authorization request parameters. This function will return an error if any
+// required parameters are missing from the claims set.
+func oidcAuthRequestFromClaims(claims map[string]any) (OIDCAuthRequestClaims, error) {
+	request := OIDCAuthRequestClaims{}
+	dec, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
+		Result: &request,
+		// Return an error if any field of OIDCAuthRequestClaims is unset
+		// with the exception of 'OptionalParameters'
+		ErrorUnset: true,
+	})
+	if err != nil {
+		return request, trace.Wrap(err, "error initializing decoder for oidc auth request claims")
+	}
+
+	err = dec.Decode(claims)
+	if err != nil {
+		return request, trace.Wrap(err, "failed to decode request object claims")
+	}
+
+	return request, err
+}
+
+// SignOIDCAuthRequest creates a JWT-Secured Authorization Request (JAR) object from the
+// authorization request.
+func (k *Key) SignOIDCAuthRequest(authRequest OIDCAuthRequestClaims) (string, error) {
+	err := authRequest.Check()
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	kid, err := KeyID(k.config.PublicKey)
+	if err != nil {
+		return "", trace.Wrap(err, "failed to determine JWT key identifier")
+	}
+
+	return k.sign(authRequest.toMap(), &jose.SignerOptions{
+		ExtraHeaders: map[jose.HeaderKey]any{
+			"kid": kid,
+		},
+	})
+}
+
+// VerifyOIDCAuthRequestToken parses and validates a JWT that is to be interpreted as
+// a JWT-Secured Authorization Request (JAR) object.
+func (k *Key) VerifyOIDCAuthRequestToken(rawToken string) (OIDCAuthRequestClaims, error) {
+	claims, err := k.verifyOnly(rawToken)
+	if err != nil {
+		return OIDCAuthRequestClaims{}, trace.Wrap(err, "error verifying authorization request object signature")
+	}
+
+	request, err := oidcAuthRequestFromClaims(claims)
+	if err != nil {
+		return OIDCAuthRequestClaims{}, trace.Wrap(err, "error validating authorization request object parameters")
+	}
+
+	return request, nil
+}
```

Comment:
if this is only used in the test maybe we should define it only in the test file rather than exporting it. Otherwise I think it should be checking the standard JWT claims like expiry, issuedAt, subject, audience, etc.